### PR TITLE
Close #1292: Added tabs for distinguishing between different users 

### LINKED ISSUE #1292: Separate identified users from anonymous users in /people
## Is your feature request related to a problem?

Currently our /people page is a very confusing list of both anonymous users with ugly UUIDs and logged in users.

## Describe the solution you'd like

I want 2 tabs (or some other solution!) where
1. users that are logged in (you can recognise them likely because they either have properties set, or because they have distinct IDs that don't look like UUIDs) and
2. anonymous users

## Describe alternatives you've considered

Manually see the difference between the two

### PR DESCRIPTION
**Pull request by yakkomajuri**
## Changes

*Please describe.*  
*If this affects the front-end, include screenshots.*  

Added 3 tabs on the 'All Users' page to allow for filtering users based on if they are anonymous (have no properties AND no distinct id) or identified. 

Thought this would be a simple frontend solution at first, but pagination was a problem, so I had to add a new query string to the `/person` endpoint. 

Added backend tests and Cypress tests. 

default tab

identified tab

anonymous tab

anonymous tab with search

## Checklist

- [x] All querysets/queries filter by Team (if this PR affects any querysets/queries)
- [x] Backend tests (if this PR affects the backend)
- [x] Cypress E2E tests (if this PR affects the front and/or backend)

### CODE REVIEW

File: cypress/plugins/index.js [hunk 1]
```diff
@@ -0,0 +1,21 @@
...
+// This function is called when a project is opened or re-opened (e.g. due to
+// the project's config changing)
+
+/**
+ * @type {Cypress.PluginConfig}
+ */
+module.exports = (on, config) => {
+  // `on` is used to hook into various events Cypress emits
+  // `config` is the resolved Cypress config
+}
```
Comment (Twixes) [inline cypress/plugins/index.js:21]: What's the point of this addition?

Comment (mariusandra) [reply cypress/plugins/index.js:21]: I think I got this file added as well one time I ran cypress tests. I didn't push it then... but might be fine to just have it in the repo.

**Comment by mariusandra** [comment]
I like the spirit of this PR! However I'm not sure the question "who is anonymous?" is easy to answer.

First, we generate a random distinct ID when `identify` is not called. However this ID is generated in the integration and can vary from one integration to the next. Here are just some IDs that I found locally:

image
image

I don't have the test data for the android/ios integrations, but I'm sure they had something different. 

You're currently using these REGEXes:

    _PROD_DISTINCT_ID_REGEX_PATTERN = r"[A-Za-z0-9]{14}-[A-Za-z0-9]{14}-[A-Za-z0-9]{8}"
    _DEMO_DISTINCT_ID_REGEX_PATTERN = r"[A-Za-z0-9]{8}-[A-Za-z0-9]{4}-[A-Za-z0-9]{4}"

We could add a bunch of other regexes here over time to cover more cases, but that leads me to point nr 2. 

These UUID distinct IDs are not "anonymous". I can absolutely imagine someone using UUIDs as primary keys in their database and calling `posthog.identify(user.id, { props })`. They would have *all* of their users then be labeled as "anonymous". It's not at all an edge case and even we are discussing (#1361) implementing similar things.

So... I don't know how to take this forward.

**Comment by timgl** [comment]
Mea culpa, I suggested this a solution in #1292 

A better solution might be to filter on whether that person has had a `$identify` event or not. That should be a reliable indicator.

File: posthog/api/person.py [hunk 2]
```diff
@@ -42,7 +42,11 @@ class CursorPagination(BaseCursorPagination):
 
 
 class PersonViewSet(viewsets.ModelViewSet):
-    renderer_classes = tuple(api_settings.DEFAULT_RENDERER_CLASSES) + (csvrenderers.PaginatedCSVRenderer,)
+    _PROD_DISTINCT_ID_REGEX_PATTERN = r"[A-Za-z0-9]{14}-[A-Za-z0-9]{14}-[A-Za-z0-9]{8}"
```
Comment (yakkomajuri) [inline posthog/api/person.py:45]: Yeah the reason I had a more "liberal" regex pattern was that on all my "anonymous" users, the hyphen after 14 alphanumeric characters was consistent, but the rest of the ID had hyphens at varying positions. Just tested this and this regex misidentifies some people for me. 

Will become irrelevant anyway though, given Tim's new suggested solution.

File: cypress/plugins/index.js (hunk 1 above)
Comment (yakkomajuri) [reply cypress/plugins/index.js:21]: Yeah I have pulled from master and gotten that file before. I guess this time I'm the one who pushed it.

**Comment by yakkomajuri** [comment]
Actually, an important question here would be: we get rid of events older than 90 days right? That should get rid of `$identify` for older users, no? I may be wrong but haven't had time to go through the backend as in-depth as I'd like just yet (weekend to-do list!).

Wondering if a more limited but still useful solution would be to filter only by the presence of properties?

**Comment by mariusandra** [comment]
We don't delete any events right now.

For properties you mean properties on the user? You can always identify the user and not pass any properties... so 🤷

**Comment by yakkomajuri** [comment]
Hmm we don't but there's a plan to do so, no? I swear I heard that somewhere. 

Either way, got you on the properties thing. Will look into optimizing via searching for the $identify event

**Comment by yakkomajuri** [comment]
Given that a Person isn't tied to their events, but rather the events are tied to a Person, I'm thinking most solutions won't be fast enough to be acceptable by a PostHog user flipping through tabs (near-instant feedback expected), once they have a significant number of events. 

Wondering if anyone has a suggestion regarding this, or if there would be opposition to adding a`is_identified` field to Person which gets set to True on an `$identify` event...

**Comment by mariusandra** [comment]
The answer to slow queries is (almost) always to precalculate the results. Thus adding a new field makes sense.

If it's the right thing to do now... or if it'll cause other problems, that I don't know. :)

**Comment by yakkomajuri** [comment]
Finally found some time to get back into this for a bit.

Switched the approach to actually tag users as identified when an "$identify" event happens. That can then quickly be referenced to sort users across tabs.

**Comment by paolodamico** [comment]
> Hmm we don't but there's a plan to do so, no? I swear I heard that somewhere.

Yes as we continue the pricing discussions it's likely that we'll introduce the data retention policies to the different plans and we'll start deleting all data as established on each plan. Might be several weeks down the road though.
CC @yakkomajuri

**Comment by yakkomajuri** [comment]
@Twixes while we're on a bit of a merge spree, do you wanna look this over so I can change what's needed and we can try to merge this today still?

You looked over the frontend before and that hasn't changed, so the focus should mostly be on the backend, since the approach changed

File: posthog/api/test/test_person.py [hunk 3]
```diff
@@ -88,6 +88,9 @@ def test_filter_by_cohort(self):
         response = self.client.get("/api/person/?cohort=%s" % cohort.pk).json()
         self.assertEqual(len(response["results"]), 1, response)
 
+    # Filters users by Identified vs Anonymous
+    # Identified users have an $identify event
+
```
Comment (Twixes) [inline posthog/api/test/test_person.py:93]: This looks like there should be a test for the `category` param here. 🤔

File: posthog/api/person.py [hunk 4]
```diff
@@ -75,6 +75,17 @@ def _filter_request(self, request: request.Request, queryset: QuerySet, team: Te
...
+        queryset_category_pass = None
+        category = request.query_params.get("category")
+        if category == "identified":
+            queryset_category_pass = queryset.filter
+        elif category == "anonymous":
+            queryset_category_pass = queryset.exclude
+        if queryset_category_pass is not None:
+            queryset = queryset_category_pass(
+                Q(is_identified=True) | ~(Q(properties={}) | Q(properties={"is_demo": True}))
+            )
```
[the remark this answers is not in the document]
Comment (Twixes) [reply posthog/api/person.py]: I think even if going the migration route, it's better to consider that in another PR. So let's get this merged. 🙂 Only left a comment regarding a seemingly missing test.

**Pull request merged**